### PR TITLE
[PREVIEW] ROC-3986 Admissions feature toggle

### PR DESCRIPTION
**Pull request by dharmendrak**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/ROC-3986

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseMapper.java [hunk 1]
```diff
@@ -92,6 +93,7 @@ public CCDCase to(Claim claim) {
             .moreTimeRequested(claim.isMoreTimeRequested() ? YES : NO)
             .claimData(claimMapper.to(claim.getClaimData()))
             .defendantEmail(claim.getDefendantEmail())
+            .features(claim.getFeatures().stream().reduce((x, y) -> x.concat(",").concat(y)).orElse(null))
```
Comment (geek-soft) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseMapper.java]: What about `.collect(Collectors.joining(", "))` to simplify it?

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseMapper.java [hunk 2]
```diff
@@ -133,7 +135,8 @@ public Claim from(CCDCase ccdCase) {
             fromNullableUTCtoLocalZone(ccdCase.getCountyCourtJudgmentRequestedAt()),
             settlement,
             fromNullableUTCtoLocalZone(ccdCase.getSettlementReachedAt()),
-            mapSealedClaimDocument(ccdCase.getSealedClaimDocument())
+            mapSealedClaimDocument(ccdCase.getSealedClaimDocument()),
+            Arrays.asList(ccdCase.getFeatures().split(","))
```
Comment (geek-soft) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseMapper.java]: Would be great to use immutable list.

REC

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/mappers/ClaimMapper.java [hunk 3]
```diff
@@ -61,6 +64,11 @@ private CountyCourtJudgment toNullableCountyCourtJudgment(String input) {
         return toNullableEntity(input, CountyCourtJudgment.class);
     }
 
+    private List<String> toNullableFeatures(String input) {
```
Comment (geek-soft) [inline ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/mappers/ClaimMapper.java]: Nothing indicates this is for features. Name could be made more generic e.g. `toList`

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/Claim.java [hunk 4]
```diff
@@ -196,6 +200,10 @@ public LocalDate getServiceDate() {
         return TotalAmountCalculator.calculateInterestForClaim(this);
     }
 
+    public List<String> getFeatures() {
```
Comment (geek-soft) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/Claim.java]: Good we do not have to check equals and hashCode anymore :) Thanks to @sabahirfan

File: src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/ClaimController.java [hunk 5]
```diff
@@ -93,10 +94,13 @@ public Claim getByClaimReference(@PathVariable("claimReference") String claimRef
...
     @ApiOperation("Creates a new claim")
-    public Claim save(@Valid @NotNull @RequestBody ClaimData claimData,
-                      @PathVariable("submitterId") String submitterId,
-                      @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation) {
-        return claimService.saveClaim(submitterId, claimData, authorisation);
+    public Claim save(
+        @Valid @NotNull @RequestBody ClaimData claimData,
+        @PathVariable("submitterId") String submitterId,
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
+        @RequestHeader("features") List<String> features
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/ClaimController.java]: Usually header names start with upper case, I would love to keep it.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ClaimServiceTest.java [hunk 6]
```diff
@@ -162,7 +162,7 @@ public void saveClaimShouldFinishSuccessfully() {
             .thenReturn(Optional.of(claim));
         when(caseRepository.saveClaim(eq(AUTHORISATION), any())).thenReturn(claim);
 
-        Claim createdClaim = claimService.saveClaim(USER_ID, claimData, AUTHORISATION);
+        Claim createdClaim = claimService.saveClaim(USER_ID, claimData, AUTHORISATION, );
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ClaimServiceTest.java]: Ups...

File: src/test/java/uk/gov/hmcts/cmc/claimstore/controllers/ClaimControllerTest.java [hunk 7]
```diff
@@ -40,7 +40,7 @@ public void setup() {
     public void shouldSaveClaimInRepository() {
         //given
         ClaimData input = SampleClaimData.validDefaults();
-        when(claimService.saveClaim(eq(USER_ID), eq(input), eq(AUTHORISATION))).thenReturn(CLAIM);
+        when(claimService.saveClaim(eq(USER_ID), eq(input), eq(AUTHORISATION), )).thenReturn(CLAIM);
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/controllers/ClaimControllerTest.java]: Ups...

File: src/main/resources/db/migration/claimstore/V2018_07_27_1348__Add_features_enabled_column.sql [hunk 8]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE claim
+  ADD COLUMN features JSON;
```
Comment (geek-soft) [inline src/main/resources/db/migration/claimstore/V2018_07_27_1348__Add_features_enabled_column.sql]: Make it JSONB please.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/FeatureTogglePermissionsController.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+@RestController
+@RequestMapping(
+    path = "/users",
+    produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+public class FeatureTogglePermissionsController {
+
+    @GetMapping("/roles")
+    @ApiOperation("Fetch user roles")
+    public String getBySubmitterId(@RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation) {
+        return "cmc-new-features-consent-given, citizen, admin, some-role";
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/FeatureTogglePermissionsController.java]: Either implement properly or remove from PR and make another PR for it.

**Comment by geek-soft** [comment]
I am happy to keep this PR focused only on adding features field to claim, storing it and retrieving it.

@dharmendrak would be good to have integration tests with headers sent to make sure it get persisted and read properly.

File: domain-sample-data/src/main/java/uk/gov/hmcts/cmc/domain/models/sampledata/SampleClaim.java [hunk 10]
```diff
@@ -102,7 +108,8 @@ public static Claim claim(ClaimData claimData, String referenceNumber) {
             null,
             null,
             null,
-            null
+            null,
+            new ArrayList<>()
```
Comment (sabahirfan) [inline domain-sample-data/src/main/java/uk/gov/hmcts/cmc/domain/models/sampledata/SampleClaim.java]: As we are passing null for many of the constructor arguments, will suggest to use builder rather than calling constructor

File: ccd-adapter/build.gradle [hunk 11]
```diff
@@ -15,6 +15,7 @@ dependencies {
   compile group: 'org.springframework.boot', name: 'spring-boot-starter', version: versions.springBoot
   annotationProcessor group: 'org.projectlombok', name: 'lombok', version: '1.16.20'
   compileOnly group: 'org.projectlombok', name: 'lombok', version: '1.16.20'
+  compile group: 'com.google.guava', name: 'guava', version: '20.0'
```
Comment (timja) [inline ccd-adapter/build.gradle]: looks unneeded?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/FeatureTogglePermissionsController.java (hunk 9 above)
Comment (timja) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/FeatureTogglePermissionsController.java]: out of scope of this PR, just move it to a separate one

File: domain-sample-data/src/main/java/uk/gov/hmcts/cmc/domain/models/sampledata/SampleClaim.java (hunk 10 above)
Comment (dharmendrak) [reply domain-sample-data/src/main/java/uk/gov/hmcts/cmc/domain/models/sampledata/SampleClaim.java]: good one Saba, added now

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/mappers/ClaimMapper.java (hunk 3 above)
Comment (dharmendrak) [reply ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/mappers/ClaimMapper.java]: done

File: ccd-adapter/build.gradle (hunk 11 above)
Comment (dharmendrak) [reply ccd-adapter/build.gradle]: ImmutableList is not available in this module

File: src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/ClaimController.java [hunk 12]
```diff
@@ -93,10 +93,13 @@ public Claim getByClaimReference(@PathVariable("claimReference") String claimRef
...
     @ApiOperation("Creates a new claim")
-    public Claim save(@Valid @NotNull @RequestBody ClaimData claimData,
-                      @PathVariable("submitterId") String submitterId,
-                      @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation) {
-        return claimService.saveClaim(submitterId, claimData, authorisation);
+    public Claim save(
+        @Valid @NotNull @RequestBody ClaimData claimData,
+        @PathVariable("submitterId") String submitterId,
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
+        @RequestHeader("Features") List<String> features
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/ClaimController.java]: I would make it optional.
